toolchain/check/context.h

```diff
@@ -252,9 +253,34 @@ class Context {
     params_or_args_stack_.AddInstId(inst_id);
   }
 
+  // Enter a declaration of kind `k`, with `f` parse node for the introducer
+  // token.
+  auto PushDeclState(DeclState::DeclKind k, Parse::Node parse_node) {
+    decl_state_stack_.push_back(DeclState(k, parse_node));
+  }
+  // Access the most recently entered declaration.
+  auto innermost_decl() -> DeclState& { return decl_state_stack_.back(); }
+  // Get the state for the declaration containing the innermost declaration.
+  // Requires that the innermost declaration is not `FileScope`.
+  auto containing_decl() const -> const DeclState& {
+    CARBON_CHECK(decl_state_stack_.size() >= 2);
+    return decl_state_stack_[decl_state_stack_.size() - 2];
+  }
+  // Leave a declaration of kind `k`.
+  auto PopDeclState(DeclState::DeclKind k) {
+    CARBON_CHECK(decl_state_stack_.back().kind == k);
+    decl_state_stack_.pop_back();
+    CARBON_CHECK(!decl_state_stack_.empty());
+  }
+
   // Prints information for a stack dump.
   auto PrintForStackDump(llvm::raw_ostream& output) const -> void;
 
+  // Get the text for the token of a node for diagnostics.
+  auto TextForNode(Parse::Node parse_node) -> llvm::StringRef {
```

Comment:
```suggestion
  auto GetTextForNode(Parse::Node parse_node) -> llvm::StringRef {
```

What would you think of adding "Get" to the function name, to clarify the action?

Reply:
I'm concerned it is verbose. I added this function specifically to make things more concise. Would `text_for_node` work?

Reply:
Maybe GetNodeText, or `node_text`?

Stepping back slightly for diagnostics, though. These are fetching the text from the source buffer from diagnostics, but I'm just thinking maybe I should be cautious about printing the raw source in diagnostics. Could they instead be passing the Lex::TokenKind and using the fixed spelling for the token?

There might also be good reasons to do that: if we get into metaprogramming, the source text might not be as expected. So far I think we've avoided having diagnostics doing that reach, instead we get things like StringifyType.

Reply:
It is now `token_kind`.